[AArch64] Define high bits of FPR and GPR registers.

sdesmalen-arm opened the pull request:
This is a step towards enabling subreg liveness tracking for AArch64, which requires that registers are fully covered by their subregisters, as covered here #109797.

There are several changes in this patch:

* AArch64RegisterInfo.td and tests: Define the high bits like B0_HI, H0_HI, S0_HI, D0_HI, Q0_HI. Because the bits must be defined by some register class, this added a register class which meant that we had to update 'magic numbers' in several tests.

  The use of ComposedSubRegIndex helped 'compress' the number of bits required for the lanemask. The correctness of the masks is tested by an explicit unit tests.

* LoadStoreOptimizer: previously 'HasDisjunctSubRegs' was only true for register tuples, but with this change to describe the high bits, a register like 'D0' will also have 'HasDisjunctSubRegs' set to true (because it's fullly covered by S0 and S0_HI). The fix here is to explicitly test if the register class is one of the known D/Q/Z tuples.

* TableGen: The handling of the isArtificial flag was entirely broken. Skipping out too early from some of the loops led to incorrect internal representation of the (sub)register(index) hierarchy, and thus resulted in incorrect TableGen info.

sdesmalen-arm committed:
[AArch64] Define high bits of FPR and GPR registers.

This is a step towards enabling subreg liveness tracking for AArch64, which
requires that registers are fully covered by their subregisters, as
covered here #109797.

There are several changes in this patch:

* AArch64RegisterInfo.td and tests: Define the high bits like B0_HI,
  H0_HI, S0_HI, D0_HI, Q0_HI. Because the bits must be defined by some
  register class, this added a register class which meant that we had
  to update 'magic numbers' in several tests.

  The use of ComposedSubRegIndex helped 'compress' the number
  of bits required for the lanemask. The correctness of the masks
  is tested by an explicit unit tests.

* LoadStoreOptimizer: previously 'HasDisjunctSubRegs' was only true for
  register tuples, but with this change to describe the high bits, a
  register like 'D0' will also have 'HasDisjunctSubRegs' set to true
  (because it's fullly covered by S0 and S0_HI). The fix here is to
  explicitly test if the register class is one of the known D/Q/Z
  tuples.

* TableGen: The handling of the isArtificial flag was entirely broken.
  Skipping out too early from some of the loops led to incorrect
  internal representation of the (sub)register(index) hierarchy, and
  thus resulted in incorrect TableGen info.

arsenm commented on llvm/unittests/Target/AArch64/AArch64RegisterInfoTest.cpp:
```diff
@@ -0,0 +1,148 @@
+
+  // Test that there is no overlap between different (sub)registers
+  // in a tuple.
+  ASSERT_EQ(TRI.getSubRegIndexLaneMask(AArch64::dsub0) &
```
EXPECT_EQ throughout

arsenm reviewed (commented):
Is it possible to split out the tablegen changes

sdesmalen-arm committed:
Fixups

sdesmalen-arm committed:
Baseline Tablegen patch, with test, no tablegen changes

sdesmalen-arm committed:
[TableGen] Fix concatenation of subreg and artificial subregs

When CoveredBySubRegs is true and a sub-register consists of two
parts; a regular subreg and an artificial subreg, then TableGen
should consider both as a concatenation of subregs.

This happens for example when a 64-bit register 'D0' consists of
32-bit 'S0_HI' (artificial) and 'S0', and 'S0' consists of (16-bit)
'H0_HI' (artificial) and 'H0'. Then the concatenation should be:
S0_HI, H0_HI, H0.

sdesmalen-arm committed:
[TableGen] Fix calculation of Lanemask for RCs with artificial subregs.

TableGen builds up a map of "SubRegIdx -> Subclass" where Subclass is
the largest class where all registers have SubRegIdx as a
sub-register. When SubRegIdx (vis-a-vis the sub-register) is
artificial it should still include it in the map. This map is used in
various places, including in the calculation of the Lanemask of a
register class, which otherwise calculates an incorrect lanemask.

sdesmalen-arm commented:
> Is it possible to split out the tablegen changes

Sure, I've created #114391 and #114392!

arsenm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -424,6 +424,57 @@ AArch64RegisterInfo::explainReservedReg(const MachineFunction &MF,
   return {};
 }
 
+static MCPhysReg ReservedHi[] = {
```
missing const

arsenm commented on llvm/lib/Target/AArch64/AArch64RegisterInfo.cpp:
```diff
@@ -424,6 +424,58 @@ AArch64RegisterInfo::explainReservedReg(const MachineFunction &MF,
   return {};
 }
 
+static SmallVector<MCPhysReg> ReservedHi = {
```
What kind of failures? AMDGPU also has synthetic 16-bit high sub registers and they are not explicitly reserved. Are you adding these to an allocatable class?

arsenm replied:
But what are the actual failures, messages, location?  If the high half of register isn't allocatable / addressable in the first place, it shouldn't just appear to cause issues

sdesmalen-arm replied:
Without marking the registers as reserved, then for the example below:
```
---
name:            sv2i64
tracksRegLiveness: true
body:             |
  bb.0.entry:
    liveins: $q0, $q1

    %0:fpr128 = COPY $q0
    %1:fpr128 = COPY $q1
    %35:gpr64 = COPY %0.dsub
    %36:gpr64 = COPY %1.dsub
    %9:gpr64 = SDIVXr %35, %36
    %37:gpr64 = UMOVvi64 %0, 1
    %38:gpr64 = UMOVvi64 %1, 1
    %10:gpr64 = SDIVXr %37, %38
    %19:fpr128 = INSvi64gpr undef %19, 0, %9
    %19:fpr128 = INSvi64gpr %19, 1, %10
    %39:gpr64 = COPY %19.dsub
    %24:gpr64 = MADDXrrr %39, %36, $xzr
    %41:gpr64 = UMOVvi64 %19, 1
    %25:gpr64 = MADDXrrr %41, %38, $xzr
    %34:fpr128 = INSvi64gpr undef %34, 0, %24
    %34:fpr128 = INSvi64gpr %34, 1, %25
    %2:fpr128 = SUBv2i64 %0, %34
    $q0 = COPY %2
    RET_ReallyLR implicit $q0
...
```

When I run this with:
```
llc -global-isel -verify-machineinstrs -run-pass=machine-scheduler
```

It fails with:
```
Use of $xzr does not have a corresponding definition on every path:
216r %10:gpr64 = MADDXrrr %9:gpr64, %3:gpr64, $xzr
LLVM ERROR: Use not jointly dominated by defs.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: ./bin/llc -global-isel -verify-machineinstrs -run-pass=machine-scheduler /tmp/t.mir -o -
1.      Running pass 'Function Pass Manager' on module '/tmp/t.mir'.
2.      Running pass 'Machine Instruction Scheduler' on function '@sv2i64'
 ...
 #8 0x0000ffff80062b7c llvm::LiveRangeCalc::findReachingDefs(llvm::LiveRange&, llvm::MachineBasicBlock&, llvm::SlotIndex, unsigned int, llvm::ArrayRef<llvm::SlotIndex>)
 #9 0x0000ffff80063e94 llvm::LiveRangeCalc::extend(llvm::LiveRange&, llvm::SlotIndex, unsigned int, llvm::ArrayRef<llvm::SlotIndex>)
#10 0x0000ffff80064a18 llvm::LiveIntervalCalc::extendToUses(llvm::LiveRange&, llvm::Register, llvm::LaneBitmask, llvm::LiveInterval*)
#11 0x0000ffff8003e82c llvm::LiveIntervals::computeRegUnitRange(llvm::LiveRange&, unsigned int)
#12 0x0000ffff80044cdc llvm::LiveIntervals::HMEditor::updateAllRanges(llvm::MachineInstr*)
#13 0x0000ffff8004848c llvm::LiveIntervals::handleMove(llvm::MachineInstr&, bool)
#14 0x0000ffff801f44ec llvm::ScheduleDAGMI::moveInstruction(llvm::MachineInstr*, llvm::MachineInstrBundleIterator<llvm::MachineInstr, false>)
#15 0x0000ffff801fdb58 llvm::ScheduleDAGMILive::scheduleMI(llvm::SUnit*, bool)
#16 0x0000ffff8020b214 llvm::ScheduleDAGMILive::schedule()
#17 0x0000ffff801f0934 (anonymous namespace)::MachineSchedulerBase::scheduleRegions(llvm::ScheduleDAGInstrs&, bool) (.isra.0) MachineScheduler.cpp:0:0
```

arsenm replied:
This smells like an unrelated bug, this is not the kind of error I expected

sdesmalen-arm replied:
I don't think there is a bug; the code for moving an instruction goes through the list of operands to update the register's liverange. For each physreg it then goes through the regunits to calculate/update the liverange for that regunit, but only if the regunit is not reserved.

The code that determines if the register is reserved says:
```
// A register unit is considered reserved if all its roots and all their
// super registers are reserved.
```
Without this change to AArch64RegisterInfo.cpp, WZR and XZR are marked as reserved, but WZR_HI isn't (because WZR_HI is a sibling of WZR, and `markSuperRegs` marks only XZR as reserved), and so `IsReserved` is `false` for the WZR_HI regunit.

Why this doesn't fail for AMDGPU I don't know, perhaps these registers are always virtual and they never go down this path.

sdesmalen-arm committed:
Add const to ReservedHi

sdesmalen-arm commented:
It wasn't, but I also didn't realise that I closed it. Could Github have done this automatically after the branch it was based of was deleted? (I was about to push the rebased branch of this PR after merging #114391 and #114392)

sdesmalen-arm commented:
Trying to reopen..

sdesmalen-arm commented:
I think I need to create a new PR for this as Github doesn't allow me to reopen and choose a different branch to merge into.
